File: backend-java/src/main/java/com/groceriesapp/service/UserService.java
package com.groceriesapp.service;

import com.groceriesapp.model.Household;
import com.groceriesapp.model.User;
import com.groceriesapp.repository.HouseholdRepository;
import com.groceriesapp.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserService {
    
    private final UserRepository userRepository;
    private final HouseholdRepository householdRepository;
    private final PasswordEncoder passwordEncoder;
    
    @Transactional
    public User createUser(String email, String password, String name, String householdName) {
        User user = new User();
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(password));
        user.setName(name);
        
        // Create or find household
        if (householdName != null && !householdName.isEmpty()) {
            Household household = new Household();
            household.setName(householdName);
            household = householdRepository.save(household);
            user.setHousehold(household);
        }
        
        return userRepository.save(user);
    }
    
    public User getUserById(Long id) {
        return userRepository.findById(id).orElse(null);
    }
    
    public User getUserByEmail(String email) {
        return userRepository.findByEmail(email).orElse(null);
    }
}

File: backend/src/main/kotlin/com/groceriesapp/service/UserService.kt
package com.groceriesapp.service

import com.groceriesapp.model.Household
import com.groceriesapp.model.User
import com.groceriesapp.repository.HouseholdRepository
import com.groceriesapp.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(
    private val userRepository: UserRepository,
    private val householdRepository: HouseholdRepository,
    private val passwordEncoder: PasswordEncoder
) {

    @Transactional
    fun createUser(email: String, password: String, name: String, householdName: String?): User {
        if (userRepository.findByEmail(email) != null) {
            throw IllegalArgumentException("User with email $email already exists.")
        }

        val household = if (householdName != null) {
            householdRepository.save(Household(name = householdName))
        } else {
            null
        }

        val newUser = User(
            email = email,
            passwordHash = passwordEncoder.encode(password),
            name = name,
            household = household
        )

        return userRepository.save(newUser)
    }

    fun findById(id: Long): User? = userRepository.findById(id).orElse(null)

    fun findByEmail(email: String): User? = userRepository.findByEmail(email)
}
